### ISSUE TITLE
[docdb] Sys catalog compaction and flush should work across peers

### ISSUE THREAD

**Issue by robertsami** [yugabyte/yugabyte-db#10948]
In https://github.com/yugabyte/yugabyte-db/issues/8730, we added commands `flush_sys_catalog` and `compact_sys_catalog`. However, these commands only work at the leader sys catalog tablet. We should enhance these commands to work across peers, not just at the leader

**Comment by SGZW** [comment]
Can I Take this issue ?

**Comment by robertsami** [comment]
@SGZW yes, please feel free to self-assign and take a look! and i encourage you to join our public slack and post in #contributors channel for more real-time feedback/input. i'll follow up in this issue with a few notes/pointers to help you tackle this issue

**Comment by robertsami** [comment]
1. see https://github.com/yugabyte/yugabyte-db/commit/3334192486fccb3dcb96eefe883ef5a1a06457c8, where we added the original functionality which compacts/flushes the sys_catalog tablet only locally
2. see https://github.com/yugabyte/yugabyte-db/blob/master/src/yb/tools/yb-admin_client.cc#L1608, where we invoke the RPC to flush the sys catalog. note that we pass in `master_admin_proxy_` to send this request to just the master leader
3. see https://github.com/yugabyte/yugabyte-db/blob/master/src/yb/tools/yb-admin_client.cc#L1543, where we actually loop over various yb-master addresses, instantiate a proxy for each and InvokeRpc to each of them

for this change, we need to do something more akin to what we do in https://github.com/yugabyte/yugabyte-db/blob/master/src/yb/tools/yb-admin_client.cc#L1543 in the compact/flush admin commands

**Comment by albertchokor** [comment]
Hi, I am a UT Austin student taking Distributed Computing, talked to Bogdan through Slack on working on this issue.

**Comment by albertchokor** [comment]
Hi @robertsami , I am a UT Austin student taking Distributed Computing. @eafdeafd and I interested in contributing to YugabyteDB for our final project.  We talked to @bmatican earlier through Slack for taking up this issue.

Thank you,
Albert

**Comment by bmatican** [comment]
@albertchokor Sounds good! Feel free to reach out to Rob, if you have any questions, else his previous comment should still be mostly applicable, although the line numbers might not match anymore.

**Comment by albertchokor** [comment]
@robertsami Could you update the line numbers in your pointers? I tried looking for them, but I could only find one I think (https://github.com/yugabyte/yugabyte-db/blob/master/src/yb/tools/yb-admin_client.cc#L1919)

**Comment by bmatican** [comment]
@albertchokor My understanding of this issue is that we send the RPC only to the master leader, but we should be sending it to all the masters. We already have that pattern in some other functions in the client, eg see https://github.com/yugabyte/yugabyte-db/blob/master/src/yb/tools/yb-admin_client.cc#L1741C41-L1741C41 (ClusterAdminClient::SetLoadBalancerEnabled), where we list all the masters, then loop through then and send the underlying RPC to each of them. In this one, we make some distinction between leader and followers, but for Flush, I think we'd just need to call the same RPC on all peers, regardless of leader vs follower.

**Comment by albertchokor** [comment]
@bmatican Taking into account what you and Robert said earlier (loop through masters, instantiate proxy, invoke RPC), @eafdeafd and I got to a solution that looks like this:
```
Status ClusterAdminClient::FlushSysCatalog() {
  // Get list of masters.
  const auto list_resp = VERIFY_RESULT(InvokeRpc(&master::MasterClusterProxy::ListMasters, *master_cluster_proxy_, ListMastersRequestPB()));

  // Instantiate proxies and invoke flush sys catalog tablet for all masters.
  master::FlushSysCatalogRequestPB req;
  for (const auto& master : list_resp.masters()) {
    HostPortPB hp_pb = master.registration().private_rpc_addresses(0);
    master::MasterAdminProxy proxy(proxy_cache_.get(), HostPortFromPB(hp_pb));
    RETURN_NOT_OK(InvokeRpc(
      &master::MasterAdminProxy::FlushSysCatalog, proxy, req));
  }

  return Status::OK();
}
```
We weren't entirely sure on how to instantiate a MasterAdminProxy based on the list we got from the MasterClusterProxy ListMasters RPC as in other examples, but this was the educated guess based on them. Please let us know if this is the step in the right direction!

Original FlushSysCatalog below:
```
Status ClusterAdminClient::FlushSysCatalog() {
  master::FlushSysCatalogRequestPB req;
  auto res = InvokeRpc(&MasterServiceProxy::FlushSysCatalog, master_proxy_.get(), req);
  return res.ok() ? Status::OK() : res.status();
}
```

Additionally, how do we update the unit test to test for multiple masters (like how to create more masters in the unit test so there is more than one + how to set follower/leader on them to test this issue):
```
TEST_F(AdminCliTest, FlushSysCatalog) {
  BuildAndStart();
  string master_address = ToString(cluster_->master()->bound_rpc_addr());
  auto client = ASSERT_RESULT(YBClientBuilder().add_master_server_addr(master_address).Build());
  ASSERT_OK(CallAdmin("flush_sys_catalog"));
}
```
